src/main/java/fi/aalto/cs/apluscourses/dal/SessionAuthentication.java

```diff
@@ -0,0 +1,86 @@
+package fi.aalto.cs.apluscourses.dal;
+
+import fi.aalto.cs.apluscourses.model.Authentication;
+import fi.aalto.cs.apluscourses.utils.Cookie;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import org.apache.http.Header;
+import org.apache.http.HttpRequest;
+import org.apache.http.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SessionAuthentication implements Authentication {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SessionAuthentication.class);
+  private final Set<Cookie> cookies = Collections.synchronizedSet(new HashSet<>());
+
+  @Override
+  public void addToRequest(HttpRequest request) {
+    URI uri = URI.create(request.getRequestLine().getUri());
+    String header = String.join(";",
+        Set.copyOf(cookies)
+            .stream()
+            .filter(cookie -> cookie.matchUri(uri))
+            .map(Cookie::toString)
+            .map(CharSequence.class::cast)
+            ::iterator);
+    LOGGER.debug(header);
+    request.addHeader("Cookie", header);
+  }
+
+  @Override
+  public boolean persist() {
+    return false;
+  }
+
+  @Override
+  public void clear() {
+
+  }
+
+  public void setCookie(Cookie cookie) {
+    LOGGER.debug("setCookie: {}", cookie);
+    synchronized (cookies) {
+      cookies.remove(cookie);
+      cookies.add(cookie);
+    }
+  }
```

Comment:
I re-implemented cookie storage only to realise that what I should have done is to re-use HttpClient object from request to request. :facepalm: :(